feat(Optimizer): Schema Inference (#1566)
nebulastream/nebulastream: docs/technical/ir_binding.md

This solves the outlined problems:

- **P1**: No prefixes are used or leaked to the user
- **P2**: One schema is one scope. In the IR, we build them per operator, but in a new binder, we can build one schema per lexical scope.
- **P3**: `LogicalFunctions` contain unambiguous references to the logical operators they were created for, allowing us to verify whether they are still up to date or update them with mappings that don't have prefix-caused edge cases.

# Field Mappings

Since we don't have source prefixes anymore, we cannot sweep the read-write conflicts in projections caused by the sequential in-place application of `MapPhysicalOperators` under the rug anymore.
Of course, we also don't want to break pipelines unnecessarily; the projections in the example query should still run in one pipeline.
But this means we cannot just materialize the new field under the intended name; we need to either redirect reads to, or writes to, the field where there is a conflict between the old and new versions.

> I found it much easier to reason about and implement redirecting the write.

Reviewer: ```suggestion
We choose to implement redirecting the write, as we found it easier to reason about.
```

Author: :+1:
